modules/keyvault/secret-autorotation/main.bicep

```diff
@@ -0,0 +1,348 @@
+  }
+}
+
+@description('Create the key rotation function app.')
+resource functions 'Microsoft.Web/sites@2022-03-01' = {
+  name: functionAppName
+  location: location
+  tags: tags
+  kind: 'functionapp'
+  identity: {
+    type: functionAppIdentityType
+    userAssignedIdentities: (functionAppIdentityType == 'UserAssigned') ? {
+      '${uai.id}': {}
+    } : null
+  }
+  properties: {
+    serverFarmId: serverfarms.id
+    httpsOnly: true
+    clientAffinityEnabled: true
+  }
+
+  resource networkConfig 'networkConfig@2022-09-01' = if (isEnableVnet) {
+    name: 'virtualNetwork'
+    properties: {
+      subnetResourceId: functionAppSubnetId
+    }
+  }
+
+  resource config 'config@2022-09-01' = {
+    name: 'appsettings'
+    properties: union({
+        AzureWebJobsStorage: 'DefaultEndpointsProtocol=https;AccountName=${appStorageAccount.name};AccountKey=${appStorageAccount.listKeys().keys[0].value};'
+        AzureWebJobsDashboard: 'DefaultEndpointsProtocol=https;AccountName=${appStorageAccount.name};AccountKey=${appStorageAccount.listKeys().keys[0].value};'
+        WEBSITE_CONTENTSHARE: functionAppName
+        WEBSITE_CONTENTAZUREFILECONNECTIONSTRING: 'DefaultEndpointsProtocol=https;AccountName=${appStorageAccount.name};AccountKey=${appStorageAccount.listKeys().keys[0].value};'
+        FUNCTIONS_EXTENSION_VERSION: '~4'
+        FUNCTIONS_WORKER_RUNTIME: 'powershell'
+        APPINSIGHTS_INSTRUMENTATIONKEY: appInsights.properties.InstrumentationKey
+        APPLICATIONINSIGHTS_CONNECTION_STRING: appInsights.properties.ConnectionString
+      }, union(isStoragePrivate ? { WEBSITE_CONTENTOVERVNET: 1 } : {},
+        functionAppIdentityType == 'SystemAssigned' ? {
+          IS_USER_ASSIGNED_IDENTITY: 'false'
+        } : {
+          IS_USER_ASSIGNED_IDENTITY: 'true'
+          USER_ASSIGNED_IDENTITY_CLIENT_ID: uai.properties.clientId
+        }))
+    dependsOn: isEnableVnet ? [ networkConfig ] : []
+  }
+}
+
+@description('Assign executor role to identity')
+module grantExecutorRole './modules/script-role-assignment.bicep' = if (isGrantExecutorRole) {
+  name: 'grant-executor-role-${suffix}'
+  params: {
+    functionAppName: functions.name
+    scriptExecutorRoleId: websiteContirbutorRoleId
+    uaiPrincipalId: uai.properties.principalId
+  }
+}
+
+@description('Deploy function app files.')
+resource deployFunctionZip 'Microsoft.Resources/deploymentScripts@2020-10-01' = {
+  name: 'upload-site-files-${suffix}'
+  location: location
+  dependsOn: isGrantExecutorRole ? [ functions, grantExecutorRole ] : [ functions ]
+  kind: 'AzureCLI'
+  identity: {
+    type: 'UserAssigned'
+    userAssignedIdentities: {
+      '${uai.id}': {}
+    }
+  }
+  properties: union({
+      azCliVersion: '2.26.1'
+      timeout: 'PT5M'
+      retentionInterval: 'PT1H'
+      scriptContent: 'echo "${zipfile}" | base64 -d > ${filename} && az functionapp deploy --resource-group ${resourceGroup().name} --name ${functions.name} --src-path ${filename} --type zip'
```

Comment:
this is short enough of a script that this is not necessary, but i prefer to pull shell scripts out into their own file, and use the loadTextContent to pull them in bicep. This lets us lint the shell scripts, and makes them a little more obvious when reviewing.

Reply:
Hi Daniel, thanks for the review and comments! We were discussing this topic earlier, a few options I think, 1) put the function source files in a publicly accessible place and use the 'sourcecontrol' sub-resource of functionapp, for example this repo itself, if it is accessible by anyone. 2) functionapp indeed supports deploying from function source code directly, the only problem is other than the function source code, there're a few shared files (e.g. host.json, profile.ps1) that needs to be uploaded to the functionapp as well, this attempt failed before but it could just be I was not doing it right.